torch/nn/functional.py

```diff
@@ -3891,6 +3843,128 @@ def multi_head_attention_forward(query,                           # type: Tensor
         assert static_v.size(2) == head_dim
         v = static_v
 
+    attn_output, attn_output_weights = scaled_dot_product_attention(
+        q, k, v, num_heads, add_zero_attn, dropout_p, training, key_padding_mask, attn_mask
+    )
+
+    assert list(attn_output.size()) == [bsz * num_heads, tgt_len, head_dim]
+
+    mha_output = multi_head_attention_out_projection(attn_output, num_heads, out_proj_weight, out_proj_bias)
+    if need_weights:
+        # average attention weights over heads
+        attn_output_weights = attn_output_weights.view(bsz, num_heads, tgt_len, -1)
+        return mha_output, attn_output_weights.sum(dim=1) / num_heads
+    else:
+        return mha_output, None
+
+
+def multi_head_attention_in_projection(query, num_heads, in_proj_weight, in_proj_bias=None):
```

Comment:
This function is same for query/key/value. We can add a single in_projection function which work on them individually. See the function [here](https://github.com/pytorch/pytorch/pull/34719/files#diff-c5db6ae353758a64f3e73dd758728cb7R378)

Reply:
That is what I have done here.